### PR TITLE
[X86][CodeGen] Add base trig intrinsic lowerings

### PR DESCRIPTION
**Pull request by farzonl**
This change is an implementation of https://github.com/llvm/llvm-project/issues/87367's investigation on supporting IEEE math operations as intrinsics.
Which was discussed in this RFC:
https://discourse.llvm.org/t/rfc-all-the-math-intrinsics/78294

This change adds constraint intrinsics and some lowering cases for `acos`, `asin`, `atan`, `cosh`, `sinh`, and `tanh`.
The only x86 specific change was for f80.

https://github.com/llvm/llvm-project/issues/70079
https://github.com/llvm/llvm-project/issues/70080
https://github.com/llvm/llvm-project/issues/70081
https://github.com/llvm/llvm-project/issues/70083
https://github.com/llvm/llvm-project/issues/70084
https://github.com/llvm/llvm-project/issues/95966
    
The x86 lowering is going to be done in three pr changes with this being the first. 
A second PR will be put up for Loop Vectorizing and then SLPVectorizer.

The constraint intrinsics is also going to be in multiple parts, but just 2. 
This part covers just the llvm specific changes, part2 will cover clang
 specifc changes and legalization for backends than have special legalization
 requirements like aarch64 and wasm.

### CODE REVIEW

**Comment by farzonl** [comment]
@RKSimon @topperc, Do you have time this week to review?

**Comment by farzonl** [comment]
@efriedma-quic I put the constraint intrinsic changes in this pr, let me know if you think it should be in a follow on pr.

File: llvm/docs/LangRef.rst [hunk 1]
```diff
@@ -26476,6 +26476,219 @@ This function returns the tangent of the specified operand, returning the
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Syntax:
+"""""""
+
+::
+
+      declare <type>
+      @llvm.experimental.constrained.asin(<type> <op1>,
+                                         metadata <rounding mode>,
```
Comment (topperc) [inline llvm/docs/LangRef.rst:26489]: I think this should be indented 1 more space

File: llvm/include/llvm/Target/TargetSelectionDAG.td [hunk 2]
```diff
@@ -1528,6 +1546,24 @@ def any_fcos       : PatFrags<(ops node:$src),
 def any_ftan       : PatFrags<(ops node:$src),
                               [(strict_ftan node:$src),
                                (ftan node:$src)]>;
+def any_fasin       : PatFrags<(ops node:$src),
+                              [(strict_fasin node:$src),
```
Comment (topperc) [inline llvm/include/llvm/Target/TargetSelectionDAG.td:1550]: Indentation is off

File: llvm/include/llvm/Target/TargetSelectionDAG.td [hunk 3]
```diff
@@ -570,6 +576,18 @@ def strict_fcos       : SDNode<"ISD::STRICT_FCOS",
                                SDTFPUnaryOp, [SDNPHasChain]>;
 def strict_ftan       : SDNode<"ISD::STRICT_FTAN",
                                SDTFPUnaryOp, [SDNPHasChain]>;
+def strict_fasin       : SDNode<"ISD::STRICT_FASIN",
+                               SDTFPUnaryOp, [SDNPHasChain]>;
```
Comment (topperc) [inline llvm/include/llvm/Target/TargetSelectionDAG.td:580]: Indentation is off

**Comment by farzonl** [comment]
@RKSimon @topperc,  @efriedma-quic  could I get another review when you get a chance?

**Comment by topperc** [comment]
This looks good to me.

**Comment by farzonl** [comment]
@efriedma-quic are you happy with the constraint intrinsic side of things? Also  @topperc thank you for the lgtm, could you also click the aprove button?

**Review by efriedma-quic — APPROVED** [review]
Didn't read carefully, but assuming the code for the new operations is doing the same thing as the code for existing operations, LGTM.

**Review by bogner — APPROVED** [review]

**Pull request merged**